src/errors.rs

```diff
@@ -35,64 +35,55 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
-use std::{io, ffi, num};
 use std::fmt::{self, Formatter, Display};
+use std::error::Error;
+use TracerState;
 
 #[derive(Debug)]
 pub enum HWTracerError {
-    // Wrapped errors from elsewhere.
-    FFIIntoString(ffi::IntoStringError),
-    FFINul(ffi::NulError),
-    IO(io::Error),
-    NumParseInt(num::ParseIntError),
-    // Our own errors.
-    CFailure,
-    HardwareSupport(String),
-    InvalidFileName(String),
-    TracerAlreadyStarted,
-    TracerDestroyed,
-    TracerNotStarted,
-    TracingNotPermitted(String),
-}
-
-impl From<ffi::IntoStringError> for HWTracerError {
-    fn from(err: ffi::IntoStringError) -> Self {
-        HWTracerError::FFIIntoString(err)
-    }
-}
+    // Non-fatal errors.
+    HWBufferOverflow,     // The trace buffer being used by the hardware overflowed.
+                          // This is considered a non-fatal error since retrying the tracing may
+                          // succeed.
+    HWSupport(String),    // The hardware doesn't support a required feature. Not fatal for the
+                          // same reason as `TracingPerms`.
+    TracingPerms(String), // Tracing is not permitted. Non-fatal as the consumer could carry on
+                          // without tracing support.
 
-impl From<ffi::NulError> for HWTracerError {
-    fn from(err: ffi::NulError) -> Self {
-        HWTracerError::FFINul(err)
-    }
+    // Fatal errors.
+    CFailure,                 // Something went wrong in C code.
+                              // ^ XXX will be replaced with an errno mechanism.
+    TracerState(TracerState), // The tracer is in the wrong state to do the requested task.
```

Comment:
I wonder whether this should just be an internal `panic!`?

Reply:
The user can easily trigger this, so I think not.

```
tracer.start_tracing();
tracer.start_tracing(); // boom
```

Reply:
Right, but since everyone will just do:

```rust
tracer.start_tracing().unwrap()
```

it seems to me that all we're doing is forcing the user to do `.unwrap()` everywhere (they're never *not* going to do `unwrap()`, as they can't recover from this error)?

Reply:
How would you feel if a library you were using as a black box crashed with a panic instead of giving you a meaningful error?

Reply:
A `panic` can give a useful error message. In this case, the onus is on the user to set up their sequence of calls to do the right thing, and it seems unlikely that they would ever do anything other than `unwrap`. Anyway, I don't feel strongly about it, so feel free to ignore me on this one.

Reply:
Another part of me also feels that I'm OK with the user crashing their own program with a hasty `unwrap()` as that clearly sends a message that the user did something wrong (starting a tracer twice). It feels somewhat rude for us to crash the user's program with something so easily triggered.